WSFrontend/org.openl.rules.ruleservice.ws/src/org/openl/rules/ruleservice/oauth/JWKInURLReader.java

```diff
@@ -0,0 +1,76 @@
+package org.openl.rules.ruleservice.oauth;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URL;
+import java.text.ParseException;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.env.Environment;
+import org.springframework.stereotype.Component;
+
+import com.nimbusds.jose.JWSHeader;
+import com.nimbusds.jose.jwk.JWK;
+import com.nimbusds.jose.jwk.JWKSet;
+import com.nimbusds.jwt.SignedJWT;
+
+/**
+ * Reader for jwk from url.
+ *
+ * @author ybiruk
+ */
+@Component
+public class JWKInURLReader implements JWKTokenReader {
+
+    private final Map<String, JWK> keyCache = new HashMap<>();
```

Comment:
Not threadsafe map implementation. Also it always growths and never cleaned. Potential memory leaks